src/contracts/actions/CappedTokenTransfers.sol

```diff
@@ -34,62 +38,126 @@ contract CappedTokenTransfers is SafeManageable, ICappedTokenTransfers {
     if (_tokens.length != _recipients.length || _tokens.length != _amounts.length) {
       revert LengthMismatch();
     }
+
     for (uint256 i = 0; i < _tokens.length; i++) {
       _addTokenTransfer(_tokens[i], _recipients[i], _amounts[i]);
     }
+
+    // Mark state as needing update
+    stateUpdated = false;
   }
 
   // ~~~ ACTIONS METHODS ~~~
 
-  function getActions() external returns (Action[] memory) {
-    Action[] memory actions = new Action[](tokenTransfers.length);
+  function getActions() external view returns (Action[] memory) {
+    // Get unique tokens and their spent amounts
+    (address[] memory uniqueTokens, uint256[] memory spentAmounts) = _getUniqueTokensAndAmounts();
+
+    // Check caps for each token
+    for (uint256 i = 0; i < uniqueTokens.length; i++) {
+      address token = uniqueTokens[i];
+      uint256 amount = spentAmounts[i];
+
+      // Validate token and cap
+      if (tokenCap[token] == 0) revert UnallowedToken();
+      if (capSpent[token] + amount > tokenCap[token]) revert ExceededCap();
+    }
 
+    // Create actions array: one for updateState + one for each transfer
+    uint256 numActions = tokenTransfers.length + 1;
+    Action[] memory actions = new Action[](numActions);
+
+    // First action: update state
+    actions[0] = Action({
+      target: address(this),
+      data: abi.encodeWithSelector(ICappedTokenTransfers.updateState.selector, uniqueTokens, spentAmounts),
+      value: 0
+    });
+
+    // Remaining actions: token transfers
     for (uint256 i = 0; i < tokenTransfers.length; i++) {
-      TokenTransfer memory tokenTransfer = tokenTransfers[i];
-      actions[i] = Action({
-        target: tokenTransfer.token,
-        data: abi.encodeWithSelector(IERC20.transfer.selector, tokenTransfer.recipient, tokenTransfer.amount),
+      TokenTransfer memory transfer = tokenTransfers[i];
+      actions[i + 1] = Action({
+        target: transfer.token,
+        data: abi.encodeWithSelector(IERC20.transfer.selector, transfer.recipient, transfer.amount),
         value: 0
       });
-      capSpent[tokenTransfer.token] += tokenTransfer.amount;
     }
 
-    for (uint256 i = 0; i < tokenTransfers.length; i++) {
-      address _token = tokenTransfers[i].token;
-      uint256 capSpentForToken = capSpent[_token];
-      if (capSpentForToken == 0) {
-        // NOTE: already processed this token
-        continue;
-      }
-      uint256 cap = tokenCap[_token];
+    return actions;
+  }
 
-      // NOTE: safety checks
-      if (cap == 0) {
-        revert UnallowedToken();
-      }
-      if (capSpentForToken > cap) {
-        revert ExceededCap();
-      }
-      if (block.timestamp < tokenCooldown[_token]) {
-        revert TokenCooldown();
-      }
+  function updateState(address[] memory _tokens, uint256[] memory _spentAmounts) external isSafe {
```

Comment:
in order to (later) make this a library, i'd suggest using `bytes memory _data` instead of custom per-contract interfaces